WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -815,6 +812,30 @@ private void LogCoinJoinSummary(ImmutableArray<AliceClient> registeredAliceClien
 			}
 		}
 
+		double winnerCost = 0;
+		while ((winner.Sum(x => x.Amount) > liquidityClue) && (winnerCost/winner.Sum(x => x.Amount) < 3))
+		{
+			List<SmartCoin> bestReducedWinner = winner;
+			var minimumAnonScore = bestReducedWinner.Min(x => x.HdPubKey.AnonymitySet);
+			var minCost = bestReducedWinner.Sum(x => (x.HdPubKey.AnonymitySet - minimumAnonScore) * x.Amount.Satoshi);
+
+			foreach (SmartCoin coin in winner.Except(new [] {selectedNonPrivateCoin}))
+			{
+				var reducedWinner = winner.Except(new [] { coin });
+				minimumAnonScore = reducedWinner.Min(x => x.HdPubKey.AnonymitySet);
+				var cost = reducedWinner.Sum(x => (x.HdPubKey.AnonymitySet - minimumAnonScore) * x.Amount.Satoshi);
+
+				if (cost < minCost)
+				{
+					minCost = cost;
+					bestReducedWinner = reducedWinner.ToList();
+				}
+			}
+
+			winner = bestReducedWinner;
+			winnerCost = minCost;
+		}
+
 		if (winner.Count != finalCandidate.Count())
 		{
 			Logger.LogDebug($"Optimizing selection, removing coins coming from the same tx.");
```

Comment:
The logging is now incorrect here. It was tailor made to the previous logic, but the added logic changes things.

Reply:
I'd suggest to move these logs up before the new code was added, then duplicate the logs here as well so we can see how the new code changed the selection.